### PR TITLE
Add support to upgrade worker nodes concurrently

### PR DESCRIPTION
**Pull request by dkoshkin**
Using `--concurrent-workers or -c` flag in`kismatic upgrade` it will upgrade multiple worker nodes concurrently.

All nodes that have `etcd` or `worker` roles will still be upgraded individually.

### CODE REVIEW

File: pkg/cli/upgrade.go [hunk 1]
```diff
@@ -52,6 +53,7 @@ Nodes in the cluster are upgraded in the following order:
 	cmd.PersistentFlags().BoolVar(&opts.skipPreflight, "skip-preflight", false, "skip upgrade pre-flight checks")
 	cmd.PersistentFlags().BoolVar(&opts.restartServices, "restart-services", false, "force restart cluster services (Use with care)")
 	cmd.PersistentFlags().BoolVar(&opts.partialAllowed, "partial-ok", false, "allow the upgrade of ready nodes, and skip nodes that have been deemed unready for upgrade")
+	cmd.PersistentFlags().IntVarP(&opts.concurrent, "concurrent-workers", "c", 1, "the number of worker nodes to upgrade concurrently")
```
Comment (alexbrand) [inline pkg/cli/upgrade.go]: I'd prefer not having a short form (-c) for this. I am also thinking this is really parallel and not concurrent.

--max-parallel-workers: The maximum number of worker nodes to be upgraded in parallel.

File: pkg/install/execute.go [hunk 2]
```diff
@@ -370,7 +370,7 @@ func (ae *ansibleExecutor) UpgradeEtcd2Nodes(plan Plan, nodesToUpgrade []Listabl
 // which phase of the upgrade we are in. For example, when upgrading a node that is both an etcd and master,
 // the etcd components and the master components will be upgraded when we are in the upgrade etcd nodes
 // phase.
-func (ae *ansibleExecutor) UpgradeNodes(plan Plan, nodesToUpgrade []ListableNode, onlineUpgrade bool) error {
+func (ae *ansibleExecutor) UpgradeNodes(plan Plan, nodesToUpgrade []ListableNode, onlineUpgrade bool, concurrent int) error {
```
Comment (alexbrand) [inline pkg/install/execute.go]: s/concurrent/maxParallelWorkers

File: pkg/install/execute.go [hunk 3]
```diff
@@ -444,6 +451,30 @@ func (ae *ansibleExecutor) upgradeNode(plan Plan, node ListableNode, onlineUpgra
 	return ae.execute(t)
 }
 
+func (ae *ansibleExecutor) upgradeNodes(plan Plan, nodes []ListableNode, onlineUpgrade bool) error {
```
Comment (alexbrand) [inline pkg/install/execute.go]: Can we make upgradeNodes a variadic functions on []ListableNode? Seems unnecessary to have both upgradeNode and upgradeNodes.

File: pkg/cli/upgrade.go [hunk 4]
```diff
@@ -105,6 +107,10 @@ before any changes are applied.
 }
 
 func doUpgrade(out io.Writer, opts *upgradeOpts) error {
+	if opts.maxParallelWorkers < 1 {
+		return fmt.Errorf("maxParallelWorkers-workers must be greater or equal to 1, got: %d", opts.maxParallelWorkers)
```
Comment (alexbrand) [inline pkg/cli/upgrade.go]: typo

File: pkg/cli/upgrade.go [hunk 5]
```diff
@@ -52,6 +53,7 @@ Nodes in the cluster are upgraded in the following order:
 	cmd.PersistentFlags().BoolVar(&opts.skipPreflight, "skip-preflight", false, "skip upgrade pre-flight checks")
 	cmd.PersistentFlags().BoolVar(&opts.restartServices, "restart-services", false, "force restart cluster services (Use with care)")
 	cmd.PersistentFlags().BoolVar(&opts.partialAllowed, "partial-ok", false, "allow the upgrade of ready nodes, and skip nodes that have been deemed unready for upgrade")
+	cmd.PersistentFlags().IntVar(&opts.maxParallelWorkers, "max-parallel-workers", 1, "the maximum number of worker nodes to be upgraded in parallel")
```
Comment (alexbrand) [inline pkg/cli/upgrade.go]: Let's move this flag to offline upgrade

**Comment by dkoshkin** [comment]
Closing to reopen again for Snap

**Pull request closed without merging**